### PR TITLE
GitHub configuration with OpenTofu

### PR DESCRIPTION
**Pull request by lmilbaum**
## Changes introduced with this PR

GitHub configuration with OpenTofu

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by jdowni000 — APPROVED** [review]
I think this is an awesome submission and I really liked the demo. As long as we address the weeds of what is exactly allowed to be auto-merged and feel comfortable with the configs I love this. This makes it extremely easy to handle the settings across the org and additional repositories added

**Review by jaredoconnell — APPROVED** [review]
This will be helpful for automating a lot of settings across all of our repositories, including those for plugins and engine components.

**Review by Harshith-umesh — COMMENTED** [review]
on reading the benefits, terraform seems to be a good addition.

**Comment by janosdebugs** [comment]
Since Terraform is no longer open source and I work on OpenTofu now, can I suggest using OpenTofu? :D

**Comment by lmilbaum** [comment]
> Since Terraform is no longer open source and I work on OpenTofu now, can I suggest using OpenTofu? :D Also, you may find some inspiration here: https://github.com/ContainerSSH/github-terraform

+1

**Comment by dustinblack** [comment]
We discussed this today in the community meeting, and it is something that we see value in pursuing. Our work on DevOps tasks for the org has been limited recently, so we haven't yet scoped in the work, but the team agrees that we should get this on the backlog even as an internal need.

We are relying on the expert opinions of @lmilbaum and @janosdebugs here since we don't otherwise have the Terraform experience, so we trust that OpenTofu may be the best way to go. Can you take the time to update the proposal to reflect this change and any corresponding details?

**Comment by lmilbaum** [comment]
> We discussed this today in the community meeting, and it is something that we see value in pursuing. Our work on DevOps tasks for the org has been limited recently, so we haven't yet scoped in the work, but the team agrees that we should get this on the backlog even as an internal need.
> 
> We are relying on the expert opinions of @lmilbaum and @janosdebugs here since we don't otherwise have the Terraform experience, so we trust that OpenTofu may be the best way to go. Can you take the time to update the proposal to reflect this change and any corresponding details?

Updated the proposal. I'll also update the code to use OpenTofu as soon as the proposal approved.

**Comment by dustinblack** [comment]
After community discussion, I want to re-open voting for this proposal and finalize whether we choose to move forward with OpenTofu. Starting from today, let's make it a 10 workin gday period for voting so that a decision can be discussed in the next community meeting on June 17th. So voting closes on Friday, June 14th.

**Review by dbutenhof — APPROVED** [review]
I actually not entirely sure how this would work in practice, much less when anyone would find bandwidth to work on it; but at least in principle having central infrastructure management of GitHub repo configuration sounds like a great idea.

**Review by dustinblack — APPROVED** [review]

**Review by webbnh — COMMENTED** [review]
I'm inclined to abstain.

Clearly, manual maintenance of ~60 repositories has been painful, and it's unlikely to get better.  And, to the extent that we expect to create additional repositories, applying IaC techniques would be key.  But, I have no idea whether OpenTofu is the "best" choice for that.

The best power of IaC is in setting up new infrastructure.  However, it seems unlikely that we will ever set up a new Arcalot on GitHub.  Far more likely is that we will need to make changes to the existing setup, and the question is, would that be best done through OpenTofu, or should we be looking to GitHub native features for that administration?

At one extreme, the Github Organization serves as a centralized locus of control.  For that, we don't need OpenTofu.  At the opposite extreme, we will presumably have a need to make changes to single repos.  Again, for that we don't need OpenTofu.  So, the times when we would need sopmething like OpenTofu are for making changes to non-trivial subsets of our repos, and I'm not sure that that is anything that we will do often enough to justify creating the requisite definitions to use in OpenTofu.

Perhaps I would think differently if I'd [seen the demo](https://github.com/arcalot/arcalot-round-table/pull/42#pullrequestreview-1665254525) or if I thought that we would have a repeated need for tweaking large numbers of repos.  However, my intuition is that, if the individual repos are set up appropriately, then we should be able to make changes at the organization-level rather than touching the individual repos.  The question then becomes, how do we straighten out the repos that we have and ensure that new repos conform to the organization requirements?  OpenTofu might be a good answer to that, or it might be a solution in search of a problem.

File: art-decisions/proposals/2023-10-09-github-configuration-as-code.md
Comment (dustinblack) [reply art-decisions/proposals/2023-10-09-github-configuration-as-code.md:3]: Your nose is always welcome. ;)

I'll check with Amye at the CNCF about the GitHub Enterprise question. If that is the case, and if we are ready to put ourselves on the CNCF sandbox track now, then perhaps that is the better approach.

**Review by dustinblack — COMMENTED** [review]
Holding my vote until I know more about the CNCF Github Enterprise question.

File: art-decisions/proposals/2023-10-09-github-configuration-as-code.md
Comment (mfleader) [reply art-decisions/proposals/2023-10-09-github-configuration-as-code.md:3]: this seems like a great idea

**Review by dustinblack — CHANGES_REQUESTED** [review]
I'm still waiting on feedback from the CNCF about the use of GitHub enterprise. We obviously aren't banking on joining the Sandbox projects, but an effort now with OpenTofu may end up redundant if we do go that route. I'm also concerned about our ability to provide resources any time soon to implement this change. So for now, I'm voting against this, but I may like to see it revisited in the future.

**Comment by webbnh** [comment]
> dustinblack [force-pushed](https://github.com/arcalot/arcalot-round-table/compare/1f337565d5b34b8009849f305d2e880b7c66cb11..d29e540593c5b23eaf6009c512a6ec4843caf652) the github-configuration branch from [1f33756 ](https://github.com/arcalot/arcalot-round-table/commit/1f337565d5b34b8009849f305d2e880b7c66cb11)to [d29e540 ](https://github.com/arcalot/arcalot-round-table/commit/d29e540593c5b23eaf6009c512a6ec4843caf652)
[12 minutes ago](https://github.com/arcalot/arcalot-round-table/pull/42#event-13104992931)

@dustinblack, what did you change?

**Review by mfleader — APPROVED** [review]

**Comment by dustinblack** [comment]
> > dustinblack [force-pushed](https://github.com/arcalot/arcalot-round-table/compare/1f337565d5b34b8009849f305d2e880b7c66cb11..d29e540593c5b23eaf6009c512a6ec4843caf652) the github-configuration branch from [1f33756 ](https://github.com/arcalot/arcalot-round-table/commit/1f337565d5b34b8009849f305d2e880b7c66cb11)to [d29e540 ](https://github.com/arcalot/arcalot-round-table/commit/d29e540593c5b23eaf6009c512a6ec4843caf652)
> > [12 minutes ago](https://github.com/arcalot/arcalot-round-table/pull/42#event-13104992931)
> 
> @dustinblack, what did you change?

Just rebased.

**Comment by webbnh** [comment]
> Just rebased.

You can do that when the PR is merged out.  Given that there are no dependencies between branches in this repo, there's no need to rebase in the middle of a review.

**Comment by dustinblack** [comment]
I've updated the help language a bit more, and I've added a default configuration so that the `-config` option is now truly optional.

I'm still not entirely satisfied. The `-context` parameter seems to have no practical function for the user. The help language implies that the workflow will be loaded from that directory, but in fact the paths for all of the inputs are relative to the current working directory.

**Review by jaredoconnell — CHANGES_REQUESTED** [review]
Given the new info, I will now reject this proposal.

**Pull request closed without merging**
